test/datacollector.PlatformTests/CommunicationLayerIntegrationTests.cs

```diff
@@ -0,0 +1,102 @@
+        private Mock<ObjectModel.Client.ITestMessageEventHandler> mockTestMessageEventHandler;
+        private string dataCollectorSettings, runSettings;
+        private IDataCollectionLauncher dataCollectionLauncher;
+
+        public CommunicationLayerIntegrationTests()
+        {
+            this.mockTestMessageEventHandler = new Mock<ObjectModel.Client.ITestMessageEventHandler>();
+            this.dataCollectorSettings = string.Format("<DataCollector friendlyName=\"CustomDataCollector\" uri=\"my://custom/datacollector\" assemblyQualifiedName=\"{0}\" codebase=\"{1}\" />", typeof(CustomDataCollector).AssemblyQualifiedName, typeof(CustomDataCollector).GetTypeInfo().Assembly.Location);
+            this.runSettings = string.Format(this.defaultRunSettings, this.dataCollectorSettings);
+#if NET46
+            this.dataCollectionLauncher = new DataCollectionLauncher();
+#else
+            this.dataCollectionLauncher = new DotnetDataCollectionLauncher();
+#endif
+        }
+
+        [TestMethod]
+        public void BeforeTestRunStartShouldGetEnviornmentVariables()
+        {
+            var dataCollectionRequestSender = new DataCollectionRequestSender();
+            var proxyDataCollectionManager = new ProxyDataCollectionManager(ObjectModel.Architecture.AnyCPU, this.runSettings, dataCollectionRequestSender, this.dataCollectionLauncher);
+
+            var result = proxyDataCollectionManager.BeforeTestRunStart(true, true, this.mockTestMessageEventHandler.Object);
+
+            Assert.AreEqual(1, result.EnvironmentVariables.Count);
+
+            proxyDataCollectionManager.Dispose();
+        }
+
+        [TestMethod]
+        public void AfterTestRunShouldSendGetAttachments()
+        {
+            var dataCollectionRequestSender = new DataCollectionRequestSender();
+            var proxyDataCollectionManager = new ProxyDataCollectionManager(ObjectModel.Architecture.AnyCPU, this.runSettings, dataCollectionRequestSender, this.dataCollectionLauncher);
+
+            proxyDataCollectionManager.BeforeTestRunStart(true, true, this.mockTestMessageEventHandler.Object);
+
+            var attachments = proxyDataCollectionManager.AfterTestRunEnd(false, this.mockTestMessageEventHandler.Object);
+
+            Assert.AreEqual("CustomDataCollector", attachments[0].DisplayName);
+            Assert.AreEqual("my://custom/datacollector", attachments[0].Uri.ToString());
+            Assert.IsTrue(attachments[0].Attachments[0].Uri.ToString().Contains("filename.txt"));
+
+            proxyDataCollectionManager.Dispose();
+        }
+
+        [TestMethod]
+        public void AfterTestRunShouldHandleSocketFailureGracefully()
+        {
+            var socketCommManager = new SocketCommunicationManager();
+            var dataCollectionRequestSender = new DataCollectionRequestSender(socketCommManager, JsonDataSerializer.Instance);
+#if NET46
```

Comment:
IDataCollectorLauncher doesn't expose Process which I am using to monitor the process.
I don't see a need to expose the process in interface, therefore I am using specific instances here.